OrbitVulkanLayer/SubmissionTrackerTest.cpp

```diff
@@ -0,0 +1,1410 @@
+              (override));
+
+  MOCK_METHOD(void, BringUp, (const std::shared_ptr<grpc::Channel>& channel), (override));
+  MOCK_METHOD(void, TakeDown, (), (override));
+
+  MOCK_METHOD(void, SetCaptureStatusListener, (CaptureStatusListener*), (override));
+
+  void StartCapture() {
+    is_capturing = true;
+    ASSERT_NE(listener_, nullptr);
+    listener_->OnCaptureStart();
+  }
+
+  void StopCapture() {
+    is_capturing = false;
+    ASSERT_NE(listener_, nullptr);
+    listener_->OnCaptureStop();
+    listener_->OnCaptureFinished();
+  }
+
+  void FakeSetCaptureStatusListener(VulkanLayerProducer::CaptureStatusListener* listener) {
+    listener_ = listener;
+  }
+
+  bool is_capturing = false;
+
+ private:
+  VulkanLayerProducer::CaptureStatusListener* listener_;
+};
+
+}  // namespace
+
+class SubmissionTrackerTest : public ::testing::Test {
+ protected:
+  void SetUp() override {
+    producer = std::make_unique<MockVulkanLayerProducer>();
+    auto set_capture_status_listener_function =
+        [this](VulkanLayerProducer::CaptureStatusListener* listener) {
+          producer->FakeSetCaptureStatusListener(listener);
+        };
+    EXPECT_CALL(*producer, SetCaptureStatusListener)
+        .Times(1)
+        .WillOnce(Invoke(set_capture_status_listener_function));
+    tracker.SetVulkanLayerProducer(producer.get());
+    auto is_capturing_function = [this]() -> bool { return producer->is_capturing; };
+    EXPECT_CALL(*producer, IsCapturing).WillRepeatedly(Invoke(is_capturing_function));
+    EXPECT_CALL(timer_query_pool, GetQueryPool).WillRepeatedly(Return(query_pool));
+    EXPECT_CALL(device_manager, GetPhysicalDeviceOfLogicalDevice)
+        .WillRepeatedly(Return(physical_device));
+    EXPECT_CALL(device_manager, GetPhysicalDeviceProperties)
+        .WillRepeatedly(Return(physical_device_properties));
+
+    EXPECT_CALL(dispatch_table, CmdWriteTimestamp)
+        .WillRepeatedly(Return(dummy_write_timestamp_function));
+  }
+
+  MockDispatchTable dispatch_table;
+  MockTimerQueryPool timer_query_pool;
+  MockDeviceManager device_manager;
+  std::unique_ptr<MockVulkanLayerProducer> producer;
+  SubmissionTracker<MockDispatchTable, MockDeviceManager, MockTimerQueryPool> tracker =
+      SubmissionTracker<MockDispatchTable, MockDeviceManager, MockTimerQueryPool>(
+          0, &dispatch_table, &timer_query_pool, &device_manager);
+
+  VkDevice device = {};
+  VkCommandPool command_pool = {};
+  VkCommandBuffer command_buffer = {};
+  VkQueryPool query_pool = {};
+  VkPhysicalDevice physical_device = {};
+  VkPhysicalDeviceProperties physical_device_properties = {.limits = {.timestampPeriod = 1.f}};
+  VkQueue queue = {};
+  VkSubmitInfo submit_info = {.sType = VK_STRUCTURE_TYPE_SUBMIT_INFO,
+                              .pNext = nullptr,
+                              .pCommandBuffers = &command_buffer,
+                              .commandBufferCount = 1};
+
```

Comment:
I guess that these should end in `_` as they're fields. Not a big fan but it's also done in the examples here: http://go/gunitprimer#same-data-multiple-tests

Reply:
Done.